File: rust/92.reverse-linked-list-ii.rs
/*
 * @lc app=leetcode id=92 lang=rust
 *
 * [92] Reverse Linked List II
 */

// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn reverse_between(
        head: Option<Box<ListNode>>,
        left: i32,
        right: i32,
    ) -> Option<Box<ListNode>> {
        let mut dummy = Some(Box::new(ListNode {
            val: -1,
            next: head,
        }));
        let mut b = &mut dummy;
        for _ in 1..left {
            if let Some(n) = b {
                b = &mut n.next;
            }
        }
        let mut p = b.as_mut().unwrap().next.take();
        let mut q = p.as_mut().unwrap().next.take();
        for _ in left..right {
            if let Some(mut n) = q {
                q = n.next;
                n.next = p;
                p = Some(n);
            }
        }
        let mut d = &mut p;
        for _ in left..right {
            if let Some(n) = d {
                d = &mut n.next;
            }
        }
        d.as_mut().unwrap().next = q;
        b.as_mut().unwrap().next = p;
        dummy.unwrap().next
    }
}
// @lc code=end

File: javascript/92.reverse-linked-list-ii.js
/*
 * @lc app=leetcode id=92 lang=javascript
 *
 * [92] Reverse Linked List II
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} left
 * @param {number} right
 * @return {ListNode}
 */
var reverseBetween = function (head, left, right) {
  const dummy = new ListNode(-1, head);
  let pre = dummy;
  for (let _ = 1; _ < left; _++) {
    pre = pre.next;
  }
  let start = pre.next,
    then = start.next;
  for (let _ = left; _ < right; _++) {
    start.next = then.next;
    then.next = pre.next;
    pre.next = then;
    then = start.next;
  }

  return dummy.next;
};
// @lc code=end
